src/plugins/Transloadit/index.js

```diff
@@ -0,0 +1,135 @@
+const Client = require('./Client')
+const StatusSocket = require('./Socket')
+
+module.exports = class Transloadit extends Plugin {
+  constructor (core, opts) {
+    super(core, opts)
+    this.type = 'uploader'
+    this.id = 'Transloadit'
+    this.title = 'Transloadit'
+
+    const defaultOptions = {
+      templateId: null,
+      wait: false
+    }
+
+    this.opts = Object.assign({}, defaultOptions, opts)
+
+    this.prepareUpload = this.prepareUpload.bind(this)
+    this.afterUpload = this.afterUpload.bind(this)
+
+    if (!this.opts.key) {
+      throw new Error('Transloadit: The `key` option is required. ' +
+        'You can find your Transloadit API key at https://transloadit.com/accounts/credentials.')
+    }
+    if (!this.opts.templateId) {
+      throw new Error('Transloadit: The `templateId` option is required. ' +
+        'You can find your template\'s ID at https://transloadit.com/templates.')
+    }
+
+    this.client = new Client({
+      key: this.opts.key
+    })
+  }
+
+  createAssembly () {
+    this.core.log('Transloadit: create assembly')
+
+    return this.client.createAssembly({
+      templateId: this.opts.templateId,
+      expectedFiles: Object.keys(this.core.state.files).length
+    }).then((assembly) => {
+      this.updateState({ assembly })
+
+      function attachAssemblyMetadata (file, assembly) {
+        const meta = Object.assign({}, file.meta, {
+          assembly_url: assembly.assembly_url,
+          filename: file.name,
+          fieldname: 'file'
+        })
+        // Add assembly-specific Tus endpoint.
+        const tus = Object.assign({}, file.tus, {
+          endpoint: assembly.tus_url
+        })
+        return Object.assign(
+          {},
+          file,
+          { meta, tus }
+        )
+      }
+
+      const filesObj = this.core.state.files
+      const files = {}
+      Object.keys(filesObj).forEach((id) => {
+        files[id] = attachAssemblyMetadata(filesObj[id], assembly)
+      })
+
+      this.core.setState({ files })
+
+      if (this.opts.wait) {
+        return this.beginWaiting()
+      }
+    }).catch((err) => {
+      this.core.emit('informer', '⚠️ Transloadit: Could not create assembly', 'error', 0)
+
+      // Reject the promise.
+      throw err
+    })
+  }
+
+  beginWaiting () {
+    this.socket = new StatusSocket(
+      this.state.assembly.websocket_url,
+      this.state.assembly
+    )
+
+    this.assemblyReady = new Promise((resolve, reject) => {
+      this.socket.on('finished', resolve)
+      this.socket.on('error', reject)
+    })
+
+    return new Promise((resolve, reject) => {
+      this.socket.on('connect', resolve)
+      this.socket.on('error', reject)
+    })
+  }
+
+  prepareUpload () {
+    this.core.emit('informer', '🔄 Preparing upload...', 'info', 0)
+    return this.createAssembly().then(() => {
+      this.core.emit('informer:hide')
+    })
+  }
+
+  afterUpload () {
+    this.core.emit('informer', '🔄 Encoding...', 'info', 0)
```

Comment:
This step can also give progress events, maybe we can integrate the pre- and postprocess steps in the dashboard progress bar to make this a bit nicer. Right now when the Encoding step starts, you see the progress bar turn green "100% uploaded" just before the "Encoding" message pops up.